File: yuizumi/galaxy.java
import java.io.*;
import java.util.*;
import java.util.regex.*;

abstract class Expr {
    public abstract Expr halfReduce();
    public abstract Expr fullReduce();
    public abstract Number asNumber();

    public Expr apply(Expr arg) {
        throw new IllegalStateException(String.format("%s is not applicable", this));
    }
}

abstract class ReducibleExpr extends Expr {
    private Expr half = null;
    private Expr full = null;

    @Override public Expr halfReduce() {
        return (half != null) ? half : (half = doHalfReduce());
    }

    @Override public Expr fullReduce() {
        return (full != null) ? full : (full = doFullReduce());
    }

    @Override public Number asNumber() {
        return fullReduce().asNumber();
    }

    protected abstract Expr doHalfReduce();
    protected abstract Expr doFullReduce();
}

abstract class GroundExpr extends Expr {
    @Override public Expr halfReduce() { return this; }
    @Override public Expr fullReduce() { return this; }

    @Override public Number asNumber() {
        throw new IllegalStateException(String.format("%s is not a number", this));
    }
}

class Number extends GroundExpr {
    private Number(long value) {
        this.value = value;
    }

    public final long value;

    public static Number of(long value) {
        return new Number(value);
    }

    @Override public Number asNumber() {
        return this;
    }

    @Override public String toString() {
        return Long.toString(value);
    }
}

class Pair extends ReducibleExpr {
    public Pair(Expr car, Expr cdr) {
        this.car = car;
        this.cdr = cdr;
    }

    public final Expr car;
    public final Expr cdr;

    @Override protected Expr doHalfReduce() {
        return this;
    }

    @Override protected Expr doFullReduce() {
        return new Pair(car.fullReduce(), cdr.fullReduce());
    }

    @Override public Expr apply(Expr x) {
        return new Apply(new Apply(x, car), cdr);
    }

    @Override public String toString() {
        return String.format("(%s, %s)", car, cdr);
    }
}

abstract class Func_1 extends GroundExpr {
    @Override public Expr apply(Expr x) {
        return invoke(x);
    }

    public abstract Expr invoke(Expr x0);
}

abstract class Func_2 extends GroundExpr {
    @Override public Expr apply(Expr x) {
        return new BindFront_2(this, x);
    }

    public abstract Expr invoke(Expr x0, Expr x1);
}

abstract class Func_3 extends GroundExpr {
    @Override public Expr apply(Expr x) {
        return new BindFront_3(this, x);
    }

    public abstract Expr invoke(Expr x0, Expr x1, Expr x2);
}

class BindFront_2 extends Func_1 {
    public BindFront_2(Func_2 fn, Expr x0) {
        this.fn = fn;
        this.x0 = x0;
    }

    public final Func_2 fn;
    public final Expr x0;

    @Override public Expr invoke(Expr x1) {
        return fn.invoke(x0, x1);
    }

    @Override public String toString() {
        return String.format("ap %s %s", fn, x0);
    }
}

class BindFront_3 extends Func_2 {
    public BindFront_3(Func_3 fn, Expr x0) {
        this.fn = fn;
        this.x0 = x0;
    }

    public final Func_3 fn;
    public final Expr x0;

    @Override public Expr invoke(Expr x1, Expr x2) {
        return fn.invoke(x0, x1, x2);
    }

    @Override public String toString() {
        return String.format("ap %s %s", fn, x0);
    }
}

// #7
class Add extends Func_2 {
    public static final Expr EXPR = new Add();

    private Add() {}

    @Override public Expr invoke(Expr x0, Expr x1) {
        return Number.of(x0.asNumber().value + x1.asNumber().value);
    }

    @Override public String toString() { return "add"; }
}

// #9
class Mul extends Func_2 {
    public static final Expr EXPR = new Mul();

    private Mul() {}

    @Override public Expr invoke(Expr x0, Expr x1) {
        return Number.of(x0.asNumber().value * x1.asNumber().value);
    }

    @Override public String toString() { return "mul"; }
}

// #10
class Div extends Func_2 {
    public static final Expr EXPR = new Div();

    private Div() {}

    @Override public Expr invoke(Expr x0, Expr x1) {
        return Number.of(x0.asNumber().value / x1.asNumber().value);
    }

    @Override public String toString() { return "div"; }
}

// #11
class Eq extends Func_2 {
    public static final Expr EXPR = new Eq();

    private Eq() {}

    @Override public Expr invoke(Expr x0, Expr x1) {
        return (x0.asNumber().value == x1.asNumber().value) ? True.EXPR : False.EXPR;
    }

    @Override public String toString() { return "eq"; }
}

// #11
class Lt extends Func_2 {
    public static final Expr EXPR = new Lt();

    private Lt() {}

    @Override public Expr invoke(Expr x0, Expr x1) {
        return (x0.asNumber().value <  x1.asNumber().value) ? True.EXPR : False.EXPR;
    }

    @Override public String toString() { return "lt"; }
}

// #16
class Neg extends Func_1 {
    public static final Expr EXPR = new Neg();

    private Neg() {}

    @Override public Expr invoke(Expr x0) {
        return Number.of(-x0.asNumber().value);
    }

    @Override public String toString() { return "neg"; }
}

// #17
class Apply extends ReducibleExpr {
    public Apply(Expr f, Expr x) {
        this.f = f;
        this.x = x;
    }

    public final Expr f;
    public final Expr x;

    @Override public Expr apply(Expr x) {
        return f.halfReduce().apply(x);
    }

    @Override protected Expr doHalfReduce() {
        return apply(x).halfReduce();
    }

    @Override protected Expr doFullReduce() {
        return apply(x).fullReduce();
    }

    @Override public String toString() {
        return String.format("ap %s %s", f, x);
    }
}

// #18
class SCombinator extends Func_3 {
    public static final Expr EXPR = new SCombinator();

    private SCombinator() {}

    @Override public Expr invoke(Expr x0, Expr x1, Expr x2) {
        return new Apply(new Apply(x0, x2), new Apply(x1, x2));
    }

    @Override public String toString() { return "s"; }
}

// #19
class CCombinator extends Func_3 {
    public static final Expr EXPR = new CCombinator();

    private CCombinator() {}

    @Override public Expr invoke(Expr x0, Expr x1, Expr x2) {
        return new Apply(new Apply(x0, x2), x1);
    }

    @Override public String toString() { return "c"; }
}

// #20
class BCombinator extends Func_3 {
    public static final Expr EXPR = new BCombinator();

    private BCombinator() {}

    @Override public Expr invoke(Expr x0, Expr x1, Expr x2) {
        return new Apply(x0, new Apply(x1, x2));
    }

    @Override public String toString() { return "b"; }
}

// #21
class True extends Func_2 {
    public static final Expr EXPR = new True();

    private True() {}

    @Override public Expr invoke(Expr x0, Expr x1) {
        return x0;
    }

    @Override public String toString() { return "t"; }
}

// #22
class False extends Func_2 {
    public static final Expr EXPR = new False();

    private False() {}

    @Override public Expr invoke(Expr x0, Expr x1) {
        return x1;
    }

    @Override public String toString() { return "f"; }
}

// #24
class ICombinator extends Func_1 {
    public static final Expr EXPR = new ICombinator();

    private ICombinator() {}

    @Override public Expr invoke(Expr x0) {
        return x0;
    }

    @Override public String toString() { return "i"; }
}

// #25
class Cons extends Func_2 {
    public static final Expr EXPR = new Cons();

    private Cons() {}

    @Override public Expr invoke(Expr x0, Expr x1) {
        return new Pair(x0, x1);
    }

    @Override public String toString() { return "cons"; }
}

// #26
class Car extends Func_1 {
    public static final Expr EXPR = new Car();

    private Car() {}

    @Override public Expr invoke(Expr x0) {
        return ((Pair) x0.halfReduce()).car;
    }

    @Override public String toString() { return "car"; }
}

// #27
class Cdr extends Func_1 {
    public static final Expr EXPR = new Cdr();

    private Cdr() {}

    @Override public Expr invoke(Expr x0) {
        return ((Pair) x0.halfReduce()).cdr;
    }

    @Override public String toString() { return "cdr"; }
}

// #28
class Nil extends Func_1 {
    public static final Expr EXPR = new Nil();

    private Nil() {}

    @Override public Expr invoke(Expr x0) {
        return True.EXPR;
    }

    @Override public String toString() { return "nil"; }
}

// #29
class IsNil extends Func_1 {
    public static final Expr EXPR = new IsNil();

    private IsNil() {}

    @Override public Expr invoke(Expr x0) {
        return (x0.fullReduce() == Nil.EXPR) ? True.EXPR : False.EXPR;
    }
}

class UserFunc extends ReducibleExpr {
    public UserFunc(String lhs, String[] rhs, Map<String, Expr> env) {
        this.lhs = lhs;
        this.rhs = rhs;
        this.env = env;
    }

    public final String lhs;
    public final String[] rhs;
    public final Map<String, Expr> env;

    private Expr compiled = null;

    private static final Pattern NUMBER = Pattern.compile("^-?[0-9]+$");

    private static final Expr AP = new Expr() {
        @Override public String toString() { return "AP"; }

        @Override public Expr halfReduce() { throw new IllegalStateException(); }
        @Override public Expr fullReduce() { throw new IllegalStateException(); }
        @Override public Expr apply(Expr arg) { throw new IllegalStateException(); }
        @Override public Number asNumber() { throw new IllegalStateException(); }
    };

    @Override protected Expr doHalfReduce() {
        return compile().halfReduce();
    }

    @Override protected Expr doFullReduce() {
        return compile().fullReduce();
    }

    private Expr tokenize(String token) {
        if (token.equals("ap")) {
            return AP;
        }
        if (env.containsKey(token)) {
            return env.get(token);
        }
        if (NUMBER.matcher(token).matches()) {
            return Number.of(Long.parseLong(token));
        }

        throw new IllegalArgumentException(token);
    }

    private Expr compile() {
        if (compiled != null) { return compiled; }

        ArrayList<Expr> stack = new ArrayList<>();

        for (String token : rhs) {
            stack.add(tokenize(token));

            while (stack.size() >= 3) {
                if (stack.get(stack.size() - 1) == AP) {
                    break;
                }
                if (stack.get(stack.size() - 2) == AP) {
                    break;
                }
                if (stack.get(stack.size() - 3) == AP) {
                    Expr x = stack.remove(stack.size() - 1);
                    Expr f = stack.remove(stack.size() - 1);
                    /* AP */ stack.remove(stack.size() - 1);
                    stack.add(new Apply(f, x));
                } else {
                    throw new IllegalArgumentException("parse error: " + lhs);
                }
            }
        }

        if (stack.size() != 1) {
            throw new IllegalArgumentException("parse error: " + lhs);
        }

        return (compiled = stack.get(0));
    }

    @Override public String toString() { return lhs; }
}

class Main {
    private static final Pattern USER_FUNC = Pattern.compile(
        "^(.*?)\\s*=\\s*(.*)$");

    public static void main(String[] args) throws Exception {
        Map<String, Expr> env = new HashMap<>();

        env.put("add", Add.EXPR);
        env.put("mul", Mul.EXPR);
        env.put("div", Div.EXPR);
        env.put("eq", Eq.EXPR);
        env.put("lt", Lt.EXPR);
        env.put("neg", Neg.EXPR);
        env.put("s", SCombinator.EXPR);
        env.put("c", CCombinator.EXPR);
        env.put("b", BCombinator.EXPR);
        env.put("i", ICombinator.EXPR);
        env.put("t", True.EXPR);
        env.put("f", False.EXPR);
        env.put("cons", Cons.EXPR);
        env.put("car", Car.EXPR);
        env.put("cdr", Cdr.EXPR);
        env.put("nil", Nil.EXPR);
        env.put("isnil", IsNil.EXPR);

        BufferedReader reader = new BufferedReader(
            new InputStreamReader(System.in));
        String line;
        while ((line = reader.readLine()) != null) {
            Matcher matcher = USER_FUNC.matcher(line);
            if (matcher.matches()) {
                String lhs = matcher.group(1);
                String rhs = matcher.group(2);
                env.put(lhs, new UserFunc(lhs, rhs.split("\\s+"), env));
            } else {
                throw new UnsupportedOperationException();
            }
        }

        Expr result = (new UserFunc("<stdin>", args, env)).fullReduce();
        System.out.println(result);
    }
}

File: portal/api/protocol/galaxy.py
import enum
import re
import sys


_NUMBER_RE = re.compile('^-?[0-9]+$')


class _Lambda(object):

    def __init__(self, name, body):
        self.name = name
        self.body = body

    def _invoke(self, arg):
        return self.body(arg)

    def __str__(self):
        return self.name


class _Special(enum.Enum):
    NIL = 'nil'
    AP = 'AP'

    def __str__(self):
        return self.value


_ENV = {
    'add'   : _Lambda('add', lambda x: _Lambda(f'(add {x})', lambda y: (_reduce(x) + _reduce(y)))),
    'mul'   : _Lambda('mul', lambda x: _Lambda(f'(mul {x})', lambda y: (_reduce(x) * _reduce(y)))),
    'div'   : _Lambda('div', lambda x: _Lambda(f'(div {x})', lambda y: (_reduce(x) // _reduce(y)))),
    'eq'    : _Lambda('eq', lambda x: _Lambda(f'(eq {x})', lambda y: (_reduce(x) == _reduce(y)))),
    'lt'    : _Lambda('lt', lambda x: _Lambda(f'(lt {x})', lambda y: (_reduce(x) < _reduce(y)))),
    'neg'   : _Lambda('neg', lambda x: -_reduce(x)),
    's'     : _Lambda('s', lambda f: _Lambda(f'(s {f})', lambda g: _Lambda(f'(s {f} {g})', lambda x: _Apply(_Apply(f, x), _Apply(g, x))))),
    'c'     : _Lambda('c', lambda f: _Lambda(f'(c {f})', lambda x: _Lambda(f'(c {f} {x})', lambda y: _Apply(_Apply(f, y), x)))),
    'b'     : _Lambda('b', lambda f: _Lambda(f'(b {f})', lambda g: _Lambda(f'(b {f} {g})', lambda x: _Apply(f, _Apply(g, x))))),
    'i'     : _Lambda('i', lambda x: x),
    't'     : True,
    'cons'  : _Lambda('cons', lambda x: _Lambda(f'(cons {x})', lambda y: _Cons(x, y))),
    'car'   : _Lambda('car', lambda x: _semi_reduce(x).car),
    'cdr'   : _Lambda('cdr', lambda x: _semi_reduce(x).cdr),
    'nil'   : _Special.NIL,
    'isnil' : _Lambda('isnil', lambda x: _reduce(x) == _Special.NIL),
}


class Error(Exception):
    pass


class _Reducible(object):

    def __init__(self):
        self._value = None

    @property
    def value(self):
        if not self._value:
            self._value = self._reduce()
        return self._value

    def _reduce(self):
        raise NotImplementedError()


class _Cons(object):

    def __init__(self, car, cdr):
        super().__init__()
        self.car = car
        self.cdr = cdr

    def __str__(self):
        return f'({self.car}, {self.cdr})'


class _Apply(_Reducible):

    def __init__(self, f, x):
        super().__init__()
        self.f = f
        self.x = x

    def _reduce(self):
        return _semi_reduce(_call(self.f, self.x))

    def __str__(self):
        return f'ap({self.f} {self.x})'


class _Expr(_Reducible):

    def __init__(self, lhs, rhs):
        super().__init__()
        self.lhs = lhs
        self.rhs = rhs

    def _reduce(self):
        stack = []

        for token in self.rhs:
            if   token == 'ap':
                stack.append(_Special.AP)
            elif token in _ENV:
                stack.append(_ENV[token])
            elif _NUMBER_RE.match(token):
                stack.append(int(token))
            else:
                raise Error(f'unknown token: {token!r}')

            while len(stack) >= 3:
                if stack[-1] == _Special.AP:
                    break
                if stack[-2] == _Special.AP:
                    break
                if stack[-3] == _Special.AP:
                    f, x = stack[-2:]
                    stack = stack[:-3]
                    stack.append(_Apply(f, x))
                else:
                    raise Error(f'invalid expression: {str(self)!r}; {stack!r}')

        if len(stack) != 1:
            raise Error(f'invalid expression: {str(self)!r}; {stack!r}')

        return _semi_reduce(stack[0])

    def __str__(self):
        return self.lhs


def _semi_reduce(x):
    return x.value if isinstance(x, _Reducible) else x


def _reduce(x):
    x = _semi_reduce(x)
    return (_reduce(x.car), _reduce(x.cdr)) if isinstance(x, _Cons) else x


def _call(f, x):
    f = _semi_reduce(f)

    if isinstance(f, _Cons):
        return _Apply(_Apply(x, f.car), f.cdr)
    if f == True:
        f = _Lambda('t', lambda x1: _Lambda(f'(t {x1})', lambda x2: x1))
    if f == False:
        f = _Lambda('f', lambda x1: _Lambda(f'(f {x1})', lambda x2: x2))
    if f == _Special.NIL:
        f = _Lambda('nil', lambda _: True)
    return f._invoke(x)


def _pretty_print(x):
    if x == _Special.NIL:
        return []
    if isinstance(x, tuple):
        car, cdr = x
        car = _pretty_print(car)
        cdr = _pretty_print(cdr)
        if isinstance(cdr, list):
            return [car] + cdr
        return (car, cdr)
    return x


def convert(arg):
    sys.setrecursionlimit(100000)
    with open("/app/api/protocol/galaxy.txt") as f:
        for line in f:
            lhs, _, *rhs = line.split()
            _ENV[lhs] = _Expr(lhs, rhs)
    return str(_pretty_print(_reduce(_Expr('<input>', arg))))


if __name__ == '__main__':
    main(sys.argv)
